fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/search/LuceneOptimizedMultiFieldQueryParser.java

```diff
@@ -72,25 +86,145 @@ protected Query getFieldQuery(String field, String queryText, int slop) throws P
+        }
+    }
+
+
+    @Override
+    @SuppressWarnings("PMD.PreserveStackTrace") //it isn't possible with Lucene's exception API
+    protected Query getRangeQuery(final String field,
+                                  final String part1,
+                                  final String part2,
+                                  final boolean startInclusive,
+                                  final boolean endInclusive) throws ParseException {
+        /*
+         * Lucene doesn't really understand types, so unless we tell it that we are looking at numeric-valued
+         * data points in our scan, it will parse everything as text, which results in incorrect results
+         * being returned over range scans.
+         *
+         * To avoid this, we use a PointConfig map (an idea taken from Lucene's StandardAnalyzer). This allows
+         * us to specify the data types for individual fields in Lucene, which we use here to create the correct
+         * type of Query object for range scans.
+         */
+        PointsConfig cfg = pointsConfig.get(field);
+        if (cfg == null) {
+            return super.getRangeQuery(field, part1, part2, startInclusive, endInclusive);
+        } else {
+            NumberFormat format = cfg.getNumberFormat();
+            Number start;
+            Number end;
+            try {
+                start = format.parse(part1);
+                end = format.parse(part2);
+            } catch (java.text.ParseException pe) {
+                throw new ParseException(QueryParserMessages.COULD_NOT_PARSE_NUMBER);
+            }
+
+            if (Integer.class.equals(cfg.getType())) {
+                int s = start.intValue();
+                int e = end.intValue();
+                //lucene range queries are inclusive, so adjust ranges as needed
+                if (!startInclusive) {
+                    s = Math.addExact(s, 1);
+                }
+                if (!endInclusive) {
+                    e = Math.addExact(e, -1);
+                }
+
+                return IntPoint.newRangeQuery(field, s, e);
+            } else if (Long.class.equals(cfg.getType())) {
+                long s = start.longValue();
+                long e = end.longValue();
+                //lucene range queries are inclusive, so adjust ranges as needed
+                if (!startInclusive) {
+                    s = Math.addExact(s, 1);
+                }
+                if (!endInclusive) {
+                    e = Math.addExact(e, -1);
+                }
+
+                return LongPoint.newRangeQuery(field, s, e);
+            } else if (Double.class.equals(cfg.getType())) {
+                double s = start.longValue();
+                double e = end.longValue();
+                //lucene range queries are inclusive, so adjust ranges as needed
+                if (!startInclusive) {
+                    s = Math.nextAfter(s, Double.MIN_VALUE);
```

Comment:
Will this make the double decrement (i.e. go in the direction of Double.MIN_VALUE)? (inclusive/exclusive is a bit odd with Double+Float)

Reply:
Doesn't his want `Double.MAX_VALUE` (with appropriate sign)? Or `POSITIVE_INFINITY` and `NEGATIVE_INFINITY`?

Reply:
according to the javadoc, `nextAfter` returns "the floating-point number adjacent to the first argument in the direction of the second argument". So if your number is x, by my read, this should basically return `x+Double.MIN_VALUE`. There are a bunch of special-case handlings for when `x == Infinity` and such, but I've adjusted the code to not go down this path for those cases.

Reply:
Ah, okay, I think you and I have a different interpretation of what "in the direction of" means. It sounds like you take it to mean that the second argument is like a signed vector from the starting point on the real line. I take it to mean that it's a second point on that line.

```
Math.nextAfter(1.0, 2.0) = 1.0000000000000002
Math.nextAfter(1.0, 0.0) = 0.9999999999999999
Math.nextAfter(1.0, Double.MIN_VALUE) = 0.9999999999999999
Math.nextAfter(1.0, 1.0) = 1.0
Math.nextAfter(1.0, Double.POSITIVE_INFINITY) = 1.0000000000000002
Math.nextAfter(1.0, Double.NEGATIVE_INFINITY) = 0.9999999999999999
Math.nextAfter(-1.0, Double.POSITIVE_INFINITY) = -0.9999999999999999
Math.nextAfter(-1.0, Double.NEGATIVE_INFINITY) = -1.0000000000000002
```

Reply:
I switched it to use `Double.MAX_VALUE` and `-Double.MAX_VALUE`